### ISSUE TITLE
Invoke failed action without triggering an Uncaught (in promise) error

### ISSUE THREAD

**Issue by milankorsos** [xdave/typescript-fsa-redux-thunk#5]
I'm trying to throw an error in my promise to dispatch the `failed` action. However if I throw an error that ends up being uncaught – what I'm trying to avoid. If I catch the exception in the promise, or in the worker then the `done` action is dispatched incorrectly. 

What is the preferred way of invoking the `failed` action while avoiding the `Uncaught (in promise) error`? 

```
const loginPromise = () => {
  return fetch('//login').then(resp => resp.json()).then(json => {
    if (!json.success) throw new Error(json.error);
    return json.success;
  });
};

const loginAction = actionCreator.async<any, any>('LOGIN');

const loginWorker = bindThunkAction(loginRequest, () => {
  return login();
});

dispatch(loginWorker());
```

Also, what is the reasoning behind re-throwing the error in `bindThunkAction`? If that'd return the error instead of throwing it that would solve my issue, though I assume it's there for a reason.

Thank you! And solid work, @xdave, keep it up!

**Comment by xdave** [comment]
The purpose was for the library to be relatively transparent to your actual code.  If an error happens to be thrown (for whatever reason, maybe a bug), I rethrow it so that you can catch it and do what needs to be done based on whatever the error is.

All this library is doing is injecting itself into the process and running a .fail() action if and when that happens.

If you think these semantics should change, I would consider it for *really* good reason. :)

**Comment by milankorsos** [comment]
I think that's a fair approach, and probably it's just me missing something. But currently I'm having problems making the above login use-case work with re-throwing the error.

1. How can I trigger the .fail() action from the promise without throwing an Uncaught promise error? Only throwing an error seems to do that, even if the promise is rejected the .done() action will be dispatched.

2. Is there a good way to catch that re-thrown error? Since the worker is just a `ThunkAction` that is passed into `dispatch()` I couldn't find a way to catch that error. I can't just wrap around `dispatch(loginWorker())` with `try .. catch`.

Thanks!

**Comment by xdave** [comment]
Promises have a .catch() method that works kind of like .then().  did you try that?

**Comment by xdave** [comment]
Apart from that, you can invoke the .failed action by dispatching it from the AsyncActionCreators that were created by the factory, ie:

store.dispatch(loginAction.failed({
  params: ..., // same params that get passed the action
  error: ..., // some error
}))

**Comment by OlafMerkert** [comment]
I was surprised, too, when I learned that the error is being rethrown.
My reasoning being: I don't want to handle the same error in two places. So if there is a "fail" action with all the error information available, I should handle these.
But as it works now, I also need to handle the error in the place I invoke the action from, and most likely I will just swallow the error there.

Would it be possible to have another `bindThunkAction` which does not rethrow the error?

**Comment by xdave** [comment]
There were two (2) main reasons for re-throwing the error:

1. It makes the Result type of the thunk unambiguous in Typescript.
    - When you specify the Result type of the thunk with `create.async<...>(...)`, your thunk must return that, and this is the type that the `done` action expects as the result.
    - This allows you to do something like `const result = await store.dispatch(someThunk(...))`, and begin using `result` in a type-safe way immediately.
    - Re-throwing the exception prevents having to specify that type as nullable or optional (ie. `SomeType | undefined`), which would then require a cast or type assertion later.
2. The logic of thunks that call other thunks will be greatly simplified in the event of errors that occur in your thunk function, especially when the result of one thunk call depends on the previous, and so on.  When the exception is thrown, execution is reverted to the catch block or the top of the stack, so instead of constantly writing something like `if (result1) { ... }` after each thunk call, you can just `await` on the responses in sequence (or in a promise chain) without all of the extra checking and type assertions.

What I could do is provide an optional third boolean argument to bindThunkAction, which, when set to false, will attempt to proceed without re-throwing the error; however, you will *need* to specify the Result type of your thunk as `SomeType | undefined` and check that result later (as the return value from a direct call, or in your reducers).

Version 2.0 is coming, in which I will also address the Result type always being a Promise (it shouldn't have to be).

**Comment by OlafMerkert** [comment]
Thanks for the clarification. I may have to extend my knowledge about thunks a bit.

**Comment by xdave** [comment]
@milankorsos @OlafMerkert  I've published 2.0 beta.  Please try it out.  It's in the `v2` branch.  Version 1.0.0 is still the default npm version for now until I'm confident that people want to use the new stuff.

**Comment by xdave** [comment]
So, I'm changing things again beyond 2.0 beta.  Besides being blocked by what I think is a bug in the new `typescript-fsa` 3.0 beta, I'm reverting back to rethrowing the error.  lol.

So the reasoning behind rethrowing the error is because the thunk functions' return type is checked.  When you specify a return type `Foo`, you must return a `Foo` from the function.  If, for some reason, an error is thrown inside of the thunk and handle some error inside of it without throwing it, the return type won't be a `Foo`.  Throwing an error is the only other option if you can't return a `Foo`.

You mentioned that the `done` action is called for the code above.  Well, I can't recall if that was happening with 1.x, but with my new code (not pushed yet), it should call the `failed` action.

Using your above code as an example:
```ts
import 'isomorphic-fetch';
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { createLogger } from 'redux-logger';
import actionCreatorFactory from 'typescript-fsa';
import { bindThunkAction } from './wtf';

const actionCreator = actionCreatorFactory('something');

const loginPromise = () => {
    return fetch('//login').then(resp => resp.json()).then(json => {
        if (!json.success) throw new Error(json.error);
        return json.success;
    });
};

const loginAction = actionCreator.async<any, any>('LOGIN');

const loginWorker = bindThunkAction(loginAction, () => {
    return loginPromise();
});

const store = createStore(
    (state: any) => state,
    applyMiddleware(thunk, createLogger({ colors: false, collapsed: true })));

store.dispatch(loginWorker({}))
    .then(success => console.log(success))
    .catch(err => err)
```
This yields (in node.js):
```
dave@davehome ~/Projects/issue-5-test % ts-node src/test
action %csomething/LOGIN_STARTED %c@ 03:24:32.926
  prev state undefined
  action     { type: 'something/LOGIN_STARTED', payload: {} }
  next state undefined
action %csomething/LOGIN_FAILED %c@ 03:24:32.956
  prev state undefined
  action     { type: 'something/LOGIN_FAILED',
    payload:
     { params: {},
       error:
        { FetchError: request to https://login failed, reason: getaddrinfo ENOTFOUND login login:443
      at ClientRequest.<anonymous> (/home/dave/Projects/issue-5-test/node_modules/node-fetch/index.js:133:11)
      at ClientRequest.emit (events.js:180:13)
      at ClientRequest.emit (domain.js:422:20)
      at TLSSocket.socketErrorListener (_http_client.js:395:9)
      at TLSSocket.emit (events.js:180:13)
      at TLSSocket.emit (domain.js:422:20)
      at emitErrorNT (internal/streams/destroy.js:64:8)
      at process._tickCallback (internal/process/next_tick.js:178:19)
          name: 'FetchError',
          message: 'request to https://login failed, reason: getaddrinfo ENOTFOUND login login:443',
          type: 'system',
          errno: 'ENOTFOUND',
          code: 'ENOTFOUND' } },
    error: true }
  next state undefined
```

This would allow you to chain several redux-thunks, depending on what you're doing:
```ts
store.dispatch(loginWorker({}))
    .then(success => store.dispatch(somethingElse(success))
    .then(more => store.dispatch(yetAnother(more))
    .catch(err => err)
```
Because the next thing in the chain depends on the result of the previous thing, execution must be interrupted when an error occurs.

**Comment by xdave** [comment]
@milankorsos @OlafMerkert OK I've found a workaround for the `typescript-fsa` bug I was talking about, and I've published a new beta (2.0.0-beta.8).  Please check it out (and the included example) and let me know if you have any issues.

**Comment by OlafMerkert** [comment]
Hi xdave, thank you for your work! Actually we replaced redux-thunk with redux-observable in our project, so I am not able to help with testing.
